zipline/utils/data.py

```diff
@@ -391,3 +399,84 @@ def _update_buffer(self, frame):
             self.buffer.loc[non_nan_items, :, non_nan_cols])
 
         self.buffer = new_buffer
+
+
+class SortedDict(MutableMapping):
+    """A mapping of key-value pairs sorted by key according to the sort_key
+    function provided to the mapping.  Ties from the sort_key are broken by
+    comparing the original keys.  `iter` traverses the keys in sort order.
+
+    Parameters
+    ----------
+    mapping : mapping, optional
+    **kwargs
+        The initial mapping.
+
+    >>> d = SortedDict(abs)
+    >>> d[-1] = 'negative one'
+    >>> d[0] = 'zero'
+    >>> d[2] = 'two'
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(0, 'zero'),
+                                         (-1, 'negative one'),
+                                         (2, 'two')])
+    >>> d[1] = 'one'  # Mutating the mapping maintains the sort order.
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(0, 'zero'),
+                                         (-1, 'negative one'),
+                                         (1, 'one'),
+                                         (2, 'two')])
+    >>> del d[0]
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one'),
+                                         (2, 'two')])
+    >>> del d[2]
+    >>> d  # doctest: +NORMALIZE_WHITESPACE
+    SortedDict[<built-in function abs>]([(-1, 'negative one'),
+                                         (1, 'one')])
+    """
+    def __init__(self, sort_key, mapping=None, **kwargs):
+        self._map = {}
+        self._sorted_key_names = []
+        self._sort_key = sort_key
+
+        self.update(merge(mapping or {}, kwargs))
+
+    def __getitem__(self, name):
+        return self._map[name]
+
+    def __setitem__(self, name, value, _bisect_right=bisect.bisect_right):
+        self._map[name] = value
+        if len(self._map) > len(self._sorted_key_names):
+            key = self._sort_key(name)
+            pair = (key, name)
+            idx = _bisect_right(self._sorted_key_names, pair)
+            self._sorted_key_names.insert(idx, pair)
+
+    def __delitem__(self, name, _bisect_right=bisect.bisect_right):
+        del self._map[name]
+        idx = _bisect_right(self._sorted_key_names,
+                            (self._sort_key(name), name))
+        del self._sorted_key_names[idx - 1]
+
+    def __iter__(self):
+        for key, name in self._sorted_key_names:
+            yield name
+
+    def __len__(self):
+        return len(self._map)
+
+    def __repr__(self, _repr_running={}):
```

Comment:
I think we are just using `_repr_running` as a `set`. We only check the keys and put in a single 1 constant. I think the intent might be more clear that way.